### PR TITLE
Updated check on model step output types from dict to Mapping

### LINKED ISSUE #17649: Loosen check on model step output types from `dict` to `Mapping`
### Description & Motivation

I want to be able to use custom classes that implement the `Mapping` interface as step outputs, and lightning should be able to work with them. `isinstance(..., dict)` is unnecessarily restrictive.

### Pitch

by example: line 63 of `loops/optimization/automatic.py` should go from this:
```
       if isinstance(training_step_output, dict):
```
to this:
```
      if isinstance(training_step_output, Mapping)
```

etc. This should (unless I am missing something) be a backward-compatible, mechanical change.

I think this change is in line with Pythonic design principles: duck typing and depending on abstractions, not concretions.

### Alternatives

_No response_

### Additional context

I wanted to use a class like this:
```
@dataclass(kw_only=True)
class HomogeneousAttrDataclass(Generic[T], Mapping[str, T]):

    def __iter__(self) -> Iterator[str]:
        return iter(it.name for it in fields(self))

    def __len__(self) -> int:
        return len(fields(self))

    def __getitem__(self, key: str) -> T:
        return getattr(self, key)
```

to be able to use nice IDE-supported dataclasses as step output types in my model, but found that lightning shuts that down with the `isinstance` check. There is a way to work around this by subclassing dict directly and then "hotwiring" the dict methods to point to dataclass fields (or whatever custom mapping impl people might want to use), but it's really ugly.

cc @borda

### PR DESCRIPTION
**Pull request by nik777**
## What does this PR do?

Fixes #17649

<details>
  <summary><b>Before submitting</b></summary>

- [x]Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18657.org.readthedocs.build/en/18657/

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
Solid! LGTM

File: src/lightning/pytorch/loops/optimization/automatic.py [hunk 1]
```diff
@@ -59,19 +59,20 @@ def _clone_loss(self) -> None:
...
+                    "the 'loss' key needs to be present"
                 )
             extra = {k: v for k, v in training_step_output.items() if k != "loss"}
         elif isinstance(training_step_output, Tensor):
             closure_loss = training_step_output
         elif training_step_output is not None:
             raise MisconfigurationException(
                 "In automatic optimization, `training_step` must return a Tensor, "
-                "a dict, or None (where the step will be skipped)."
+                "a Mapping, or None (where the step will be skipped)."
```
Comment (awaelchli) [inline src/lightning/pytorch/loops/optimization/automatic.py:75]:
```suggestion
                "a Mapping (dict), or None (where the step will be skipped)."
```
(a beginner might not know hat a mapping is)

**Review by carmocca — APPROVED** [review]
I would still use "dict" in the error messages. Most people don't understand the difference between them.
